File: CreationPattern01_AbstractFactory/cpp_src/wolf_attack_story.cpp
/**
 * @file wolf_attack_story.cpp
 * @author PFA03027@nifty.com
 * @brief implementation of wolf attack story
 * @version 0.1
 * @date 2022-07-02
 * 
 * @copyright Copyright (c) 2022, PFA03027@nifty.com
 * 
 */


#include "wolf_attack_story.h"

/**
 * @brief 家を立てた後に、オオカミが攻撃してくるというお話
 * 
 * @param p_hf 
 */
void wolf_attack_story(abstract_housemaker_factory* p_hf )
{
    std::unique_ptr<abstract_house_product> up_house_obj = p_hf->make_house();
    std::unique_ptr<abstract_garden_product> up_garden_obj = p_hf->make_garden();

    up_house_obj->what_i_am();
    up_garden_obj->what_i_am();

    up_house_obj->wolf_attacked();

    return;
}

File: CreationPattern01_AbstractFactory/rust_src/src/wolf_attack_story.rs
/// このモジュールは、AbstractHousemakerFactoryトレイトのみに依存する。
/// よって、実際の構造体定義を知る必要はないため、useによる参照を含まない。
/// 
/// これにより、具象クラスが変更されても、このモジュールは再利用可能となり、Abstract Factoryパターンの意図が実現されたことととなる。
/// 

use crate::abstract_factory_housemaker::AbstractHousemakerFactory;

/// 家を立てた後に、オオカミが攻撃してくるというお話
pub fn wolf_attack_story(hm: Box<dyn AbstractHousemakerFactory>) {
    let built_house = hm.make_house();
    let built_garden = hm.make_garden();

    built_house.what_i_am();
    built_garden.what_i_am();
    built_house.wolf_attacked();
}
